src/main/java/com/google/api/generator/gapic/composer/common/AbstractServiceSettingsClassComposer.java

```diff
@@ -75,22 +74,25 @@
 import java.util.stream.Collectors;
 import javax.annotation.Generated;
 
-public class ServiceSettingsClassComposer implements ClassComposer {
+public abstract class AbstractServiceSettingsClassComposer implements ClassComposer {
   private static final String BUILDER_CLASS_NAME = "Builder";
   private static final String CALL_SETTINGS_TYPE_NAME_PATTERN = "%sCallSettings";
   private static final String PAGED_RESPONSE_TYPE_NAME_PATTERN = "%sPagedResponse";
 
   private static final String OPERATION_SETTINGS_LITERAL = "OperationSettings";
   private static final String SETTINGS_LITERAL = "Settings";
 
-  private static final ServiceSettingsClassComposer INSTANCE = new ServiceSettingsClassComposer();
-
-  private static final TypeStore FIXED_TYPESTORE = createStaticTypes();
-
-  private ServiceSettingsClassComposer() {}
-
-  public static ServiceSettingsClassComposer instance() {
-    return INSTANCE;
+  private final TypeStore fixedTypeStore;
+  private final Class<?> instantiatingChannelProviderClass;
+  private final String defaultTransportProviderBuilderName;
+
+  protected AbstractServiceSettingsClassComposer(
+      TypeStore fixedTypeStore,
+      Class<?> instantiatingChannelProviderClass,
+      String defaultTransportProviderBuilderName) {
+    this.fixedTypeStore = fixedTypeStore;
+    this.instantiatingChannelProviderClass = instantiatingChannelProviderClass;
+    this.defaultTransportProviderBuilderName = defaultTransportProviderBuilderName;
```

Comment:
Responding to [this comment](https://github.com/googleapis/gapic-generator-java/pull/717#discussion_r618675308):

To recap, we currently have three possible approaches:
- Instantiate any transport-specific values in the transport composer (e.g. GrpcServiceSettingsClassComposer sets `defaultTransportProviderBuilderName = "defaultGrpcTransportProviderBuilder"` (current approach in this PR).
  - Advantages: Maximizes per-composer locality.
  - Drawbacks: Splits per-transport state across many composers, so it may be harder to debug or change transport-specific things. When `ClassNames` is later involved, there will be another potential set of classes to look up.
- Pass the Transport enum into the Abstract class, which can then decide whether to use `"Grpc"` or `"HttpJson"` in the string (**most suboptimal**).
  - Advantages: Maximizes locality in the Abstract class's common logic (of which there's a significant amount).
  - Drawbacks: The abstract class should not know anything about per-transport state. 
- Have a per-transport context object that holds the respective values. 
  - Advantages: Maximizes per-transport locality, transport-specific debugging may be easier.
  - Drawbacks: Need to look up context in a transport-specific file.

My $0.02: The per-transport context object doesn't have to hold all the state, just anything that needed to be passed into the superclass. Motivation: Passing in differing string values in each per-transport composer class may possibly be less maintainable, especially since the values change within the same transport (e.g. `"gRPC"` vs `"Grpc"`).

We can also discuss more offline. :)

Reply:
I can do the contexts, no problem, two questions before doing it (we can discuss it via GVC, just mentioning it here to not forget it):

Please clarify what you mean by context maximizing per-transport locality?

Yes, the context will need to hold the stuff to be passed to the superclass, but there are multiple superlcasses (like 5+) and each of them needs a different set of things (with some intersection, but not much). This means we will either need 5+ different transport specific contexts, or one transport-specific context with stuff shared across different composers in, which will tightly couple the composers class hierarchy with the context and with each other.

Reply:
Discussed offline.